### PR TITLE
Implement support for cleartext authentication plugin

### PR DESCRIPTION
**Pull request by wildmaples**
Resolves: https://github.com/trilogy-libraries/trilogy/issues/157
Redo of: https://github.com/trilogy-libraries/trilogy/pull/168

The server doesn't request for cleartext plugin in the first handshake as it is not a default plugin. The client should only provide cleartext password when the server requests an auth switch to it, and the client has `enable_cleartext_plugin` option enabled. If the client does not have cleartext auth plugin enabled, we will raise a `Trilogy::AuthPluginError`. 

This is the authentication flow for a server that uses the cleartext client side plugin (https://github.com/trilogy-libraries/trilogy/pull/168#discussion_r1548811735): 
> Server sends handshake, with auth plugin caching_sha2_password
> Client sends handshake response, still with auth plugin caching_sha2_password
> Server sends auth switch request, now with auth plugin mysql_clear_password
> Client sends auth switch response with just the cleartext password and nothing else
> Server sends OK and we are connected

This PR introduces these changes: 
- Supports cleartext auth plugin! 
- Raises when the server auth switches to a plugin that Trilogy does not support. Note that at the moment Trilogy supports mysql_native_password, caching_sha2_password and now mysql_clear_password. This is the recommended behaviour according to mysql [docs "Authentication Method Mismatch"](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase.html#sect_protocol_connection_phase_auth_method_mismatch).
- Adds a test dependency on this [mysql test server plugin](https://github.com/mysql/mysql-server/blob/824e2b4064053f7daf17d7f3f84b7a3ed92e5fb4/plugin/auth/test_plugin.cc#L176-L217). It's unclear if the file will stick around in the long term but seeing that it is only a test dependency and makes our lives easier than the alternative testing options, we are inclined to just use it for the time being. 

cc: @ngan @composerinteralia

### CODE REVIEW

File: src/protocol.c
Comment (composerinteralia) [inline src/protocol.c:619]:
```suggestion
            } else if (!strcmp("mysql_native_password", auth_plugin)) {
```

**Comment by ngan** [comment]
I created an issue on the MySQL docker repo to see if they'd be open to building and tagging a `mysql-community-test` variant that we can use. If they do, we can switch to that and get the test auth plugin for free. If they don't, we can keep what we have (checked-in plugins for various mysql version), or build our own MySQL image.
 
https://github.com/docker-library/mysql/issues/1040

**Review by composerinteralia — APPROVED** [review]
This looks great to me, but I'll leave it for a bit in case anybody else has feedback.

File: contrib/ruby/ext/trilogy-ruby/cext.c
Comment (casperisfine) [inline contrib/ruby/ext/trilogy-ruby/cext.c:1214]:
```suggestion
    rb_global_variable(&Trilogy_AuthPluginError);
    Trilogy_AuthPluginError = rb_const_get(Trilogy, rb_intern("AuthPluginError"));
```

Global variables should be registered before they are assigned. The windows for a bug is extremely small, so most gems get away with it, but we should correct that.

File: test/auth_plugins/5.7/auth_test_plugin.so
Comment (casperisfine) [inline test/auth_plugins/5.7/auth_test_plugin.so:1]: This probably shouldn't be committed?

Comment (adrianna-chang-shopify) [reply test/auth_plugins/5.7/auth_test_plugin.so:1]: Should we be building our own MySQL image instead of committing the plugins?

**Review by adrianna-chang-shopify — COMMENTED** [review]
Thanks for working on this! Would you mind squashing your commits?

My preference would be for us to build our own MySQL image extending `mysql-community-test` over checking in those plugins, but don't feel strongly if folks would prefer to ship as is.

**Comment by composerinteralia** [comment]
> Would you mind squashing your commits?

I do like a clean git history, but we haven't historically been as opinionated about this as (for example) Rails.

File: test/auth_plugins/5.7/auth_test_plugin.so
Comment (ngan) [reply test/auth_plugins/5.7/auth_test_plugin.so:1]: I think we can and should, but that would be quite a big detour from the original intent of this PR. I suggest we land as-is then I can spend some time refactoring the test setup.

Comment (adrianna-chang-shopify) [reply test/auth_plugins/5.7/auth_test_plugin.so:1]: Follow up sounds good to me 👍 Thanks!

**Review by adrianna-chang-shopify — APPROVED** [review]
> I do like a clean git history, but we haven't historically been as opinionated about this as (for example) Rails.

I'll defer to you folks, I wasn't sure if we were following Rails' stance of squash-everything-before-merge! 🙇‍♀️

File: contrib/ruby/test/auth_test.rb
Comment (ngan) [reply contrib/ruby/test/auth_test.rb:3]: Done!

Comment (ngan) [reply contrib/ruby/test/auth_test.rb:3]: Actually, I could only add it for native password. We don't use cache_sha2 for mysql 5 so the ruby test fails there. I'm going to revisit this after this PR.

Comment (composerinteralia) [reply contrib/ruby/test/auth_test.rb:3]: I think we already have that test elsewhere, but we can sort it out later. We might want to move all connection-related stuff into something like ConnectionTest.

**Comment by composerinteralia** [comment]
Thank you!

File: test/auth_plugins/5.7/auth_test_plugin.so
Comment (ngan) [reply test/auth_plugins/5.7/auth_test_plugin.so:1]: This has been addressed here: https://github.com/trilogy-libraries/trilogy/pull/174

**Pull request merged**
